### PR TITLE
add lotka volterra

### PR DESCRIPTION
**Pull request by Justinezgh**

### CODE REVIEW

File: sbids/tasks/lotkavolterra.py [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+from jax.experimental.ode import odeint
+from tensorflow_probability.substrates import jax as tfp
+tfd = tfp.distributions
+tfb = tfp.bijectors
+
+
+class Lotka_Volterra:
+
+    def __init__(self,
+                num_seed:int=None):
```
Comment (EiffL) [inline sbids/tasks/lotkavolterra.py:13]: This is not good design for several reasons:
1. If a num_seed is necessary, then it shouldn't be left with None default. As a user, the first thing I tried was to instantiate the model like this "LV = Lotka_Volterra(); LV.get_truth()", which only at the second step triggered an error complaining that seed cannot be None.
2. In JAX code, each random function should explicitly be a function of an rng key. You do not want to have separate pieces of the code instantiate their own "jax.random.PRNGKey(self.num_seed)". This should happen only once, at the beginning of the JAX script, and all subsequent random function calls should use a key derived from that top key

File: sbids/tasks/lotkavolterra.py [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+
+            dX_dt = alpha * X - beta * X * Y
+            dY_dt = -gamma * Y + delta * X * Y
+            return jnp.stack([dX_dt, dY_dt],axis=0)
+        self.equation = equation
+    
+        @jax.jit
+        def solve_equation(init,theta):
+          
+            ts = jnp.arange(0.,10.,0.1)
```
Comment (EiffL) [inline sbids/tasks/lotkavolterra.py:49]: humm that looks like some hard coded things, I would say these need to be an argument of the simulator

File: sbids/tasks/lotkavolterra.py [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+            z = odeint(equation, init, ts, theta, rtol=1e-9, atol=1e-9)
```
Comment (EiffL) [inline sbids/tasks/lotkavolterra.py:50]: 1e-9 is VERY strict, the numerical accuracy of floating point computations is 1e-7, you don't want to ask for an rtol smaller than this. I'd say 1e-5 should already be plenty

File: sbids/tasks/lotkavolterra.py [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+            z = z.T[:,::21]
```
Comment (EiffL) [inline sbids/tasks/lotkavolterra.py:51]: why only one point out of 21? That seems arbitrary, and probably dependent of the inpt t

File: sbids/tasks/lotkavolterra.py [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+
+            return z
+        self.solve_equation = solve_equation
+
+
+    def get_truth(self):
```
Comment (EiffL) [inline sbids/tasks/lotkavolterra.py:57]: this function should be function of an rng key, that gets splitted.

Comment (EiffL) [reply sbids/tasks/lotkavolterra.py:57]: and actually, I think there should be no difference between simulation and truth. It's just a draw from from the probabilistic model.

**Review by EiffL — CHANGES_REQUESTED** [review]

**Comment by EiffL** [comment]
I made some general comments, I think there is a bit too much code repetition, which is dangerous, and I actually ran into an error on my first try at trying to use it ^^'

**Comment by EiffL** [comment]
Small note ^^
Check that out:
https://num.pyro.ai/en/stable/examples/ode.html

File: sbids/tasks/lotkavolterra.py [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+            Dataset (theta, observation, score).
+            'Theta': the model parameters.
+            'Observation': the simulations from p(x|theta). 
+            'Score': the joint score.
+        """
+
+        key1, key2, key3, key4 = jax.random.split(seed,4)
+
+        prior_parameters = tfd.Independent(tfd.LogNormal(jnp.array([-0.125,-3,-0.125,-3]), 0.5*jnp.ones(4)),1)
+        prior_latent_variable = tfd.LogNormal(jnp.log(10*jnp.ones(2)), 0.8*jnp.ones(2))
```
Comment (EiffL) [inline sbids/tasks/lotkavolterra.py:89]: case and point of why it's very bad practice to duplicate code >.< the prior on latents here is different from the one in the `get_reference_posterior` function which is:
```
tfd.LogNormal(jnp.log(jnp.ones(2)*3), 0.5*jnp.ones(2))
```
